ow_sim_tests/test/terrain_interaction.py

```diff
@@ -0,0 +1,329 @@
+#!/usr/bin/env python
+
+# The Notices and Disclaimers for Ocean Worlds Autonomy Testbed for Exploration
+# Research and Simulation can be found in README.md in the root directory of
+# this repository.
+
+from math import sqrt
+from copy import copy
+import unittest
+import rospy
+import roslib
+import actionlib
+
+from gazebo_msgs.msg import ModelStates, LinkStates, LinkState
+from geometry_msgs.msg import Point
+
+import ow_lander.msg
+
+PKG = 'ow_sim_tests'
+roslib.load_manifest(PKG)
+
+GROUND_POSITION = -0.155
+
+SCOOP_LINK_NAME = 'lander::l_scoop'
+SAMPLE_DOCK_LINK_NAME = 'lander::lander_sample_dock_link'
+
+REGOLITH_PREFIX = 'regolith_'
+
+"""
+Computes the 3D distance between two geometry_msgs.msg Points
+"""
+def distance(p1, p2):
+  v = Point(p2.x - p1.x,
+            p2.y - p1.y,
+            p2.z - p1.z)
+  return sqrt(v.x*v.x + v.y*v.y + v.z*v.z)
+
+"""
+Computes the distance in only the X-Y plane between two geometry_msgs.msg Points
+"""
+def distance_flat_xy(p1, p2):
+  v = Point(p2.x - p1.x,
+            p2.y - p1.y,
+            0)
+  return sqrt(v.x*v.x + v.y*v.y)
+
+class TerrainInteraction(unittest.TestCase):
+
+  def __init__(self, *args, **kwargs):
+    unittest.TestCase.__init__(self, *args, **kwargs)
+    rospy.init_node('terrain_interaction_test', anonymous=True)
+
+    # subscribe to get all gazebo link positions
+    self._gz_link_names = []
+    self._gz_link_poses = []
+    self._link_states_sub = rospy.Subscriber("/gazebo/link_states",
+                                             LinkStates,
+                                             self._on_link_states)
+
+    # proceed with test only when ros clock has been initialized
+    while rospy.get_time() == 0:
+      rospy.sleep(0.1)
+
+  """
+  Callback function for the link_states topic
+  """
+  def _on_link_states(self, message):
+    self._gz_link_names = message.name
+    self._gz_link_poses = message.pose
+
+  """
+  Returns true if an action is done
+  @param action_client: ROS action client object
+  """
+  def _is_action_done(self, action_client):
+    return action_client.simple_state == actionlib.SimpleGoalState.DONE
+
+  """
+  Returns true if model/link name matches the regolith prefix
+  @param name: Name of a model/link
+  """
+  def _is_regolith(self, name):
+    return name[:len(REGOLITH_PREFIX)] == REGOLITH_PREFIX
+
+  """
+  Returns the position of a givne link
+  @param link_name: Name of a valid gazebo link
+  """
+  def _get_link_position(self, link_name):
+    try:
+      i = self._gz_link_names.index(link_name)
+    except ValueError as e:
+      rospy.logwarn(e)
+      return None
+    else:
+      # make a shallow copy
+      # otherwise functions that call this can modify positions by reference
+      return copy(self._gz_link_poses[i].position)
```

Comment:
with regard to the **copy**, nothing wrong with copying the value but I am not sure what this matters. After all there are many ways you can abuse variables and member in python if someone wish to.

Reply:
It is less about restricting a caller of `_get_link_positions` from modifying the member list, and more about making the getter behave in the way I typically think of getters working to avoid a pitfall which I fell into while developing this script.

Before I added the `distance_flat_xy` function, there was a portion of the code that modified the values of a returned link position. Since it returned a mutable reference, the actual values of `self._gz_link_poses` were being modified, and that oversight resulted in a debugging headache which I would like to spare anyone else that happens to work on this code, including myself in case I forget the dangers of python referencing in the future.